### PR TITLE
Implement api endpoint for logs.

### PR DESCRIPTION
**Pull request by hibooboo2**

### CODE REVIEW

File: code/iaas/model/src/main/resources/META-INF/cattle/core-model/spring-core-model-context.xml [hunk 1]
```diff
@@ -33,6 +33,7 @@
             <set>
                 <value>virtualMachine,parent=instance</value>
                 <value>container,parent=instance</value>
+                <value>containerLogsInput</value>
```
Comment (ibuildthecloud) [inline code/iaas/model/src/main/resources/META-INF/cattle/core-model/spring-core-model-context.xml]: You should be registering the class ContainerLogs, not the string type.  Add it at https://github.com/rancherio/cattle/pull/210/files#diff-c619be8811e6d62c27a8064d172fb143R19

Comment (hibooboo2) [inline code/iaas/model/src/main/resources/META-INF/cattle/core-model/spring-core-model-context.xml]: ok

File: resources/content/schema/base/containerLogsInput.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
+{
+    "collectionMethods":[],
+    "resourceFields":{
+        "follow":{
```
Comment (ibuildthecloud) [inline resources/content/schema/base/containerLogsInput.json]: Set nullable=false and default to true

File: resources/content/schema/base/containerLogsInput.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+            "type":"boolean",
+            "nullable":true,
+            "default":"true"
+        },
+        "lines":{
+            "type":"int",
+            "nullable":true,
+            "default":"100"
```
Comment (ibuildthecloud) [inline resources/content/schema/base/containerLogsInput.json]: set nullable to false and default to 100, not "100"

File: resources/content/schema/user/user-auth.json [hunk 4]
```diff
@@ -64,10 +64,10 @@
         "containerExec.attachStdout" : "cr",
         "containerExec.command" : "cr",
         "containerExec.tty" : "cr",
-
-        "containerExecOutput" : "r",
-        "containerExecOutput.token" : "r",
-        "containerExecOutput.url" : "r",
+        
+        "containerLogsInput":"cr",
```
Comment (ibuildthecloud) [inline resources/content/schema/user/user-auth.json]: This type shouldn't exist, the type should be "containerLogs"

File: resources/content/schema/user/user-auth.json [hunk 5]
```diff
@@ -195,7 +199,7 @@
         "loadBalancerTarget.instanceId" : "r",
         "loadBalancerTarget.loadBalancerId" : "r",
         "loadBalancerTarget.ipAddress" : "r",
-
+
```
Comment (ibuildthecloud) [inline resources/content/schema/user/user-auth.json]: whitespace

File: code/implementation/docker/api/src/main/java/io/cattle/platform/docker/api/ContainerLogsActionHandler.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            host = DockerUtils.getHostFromContainer(objectManager, container, null);
+        }
+
+        if (host == null) {
+            return null;
+        }
+
+        ContainerLogs exec = request.proxyRequestObject(ContainerLogs.class);
+
+        Map<String, Object> data = CollectionUtils.asMap(DockerInstanceConstants.DOCKER_ATTACH_STDIN, true, DockerInstanceConstants.DOCKER_ATTACH_STDOUT, true,
```
Comment (ibuildthecloud) [inline code/implementation/docker/api/src/main/java/io/cattle/platform/docker/api/ContainerLogsActionHandler.java]: Such ugly formatting...

Comment (ibuildthecloud) [inline code/implementation/docker/api/src/main/java/io/cattle/platform/docker/api/ContainerLogsActionHandler.java]: Most of these constants are needed. ATTACH_STDIN/STDOUT.  TTY makes no sense.  CMD???

Comment (hibooboo2) [inline code/implementation/docker/api/src/main/java/io/cattle/platform/docker/api/ContainerLogsActionHandler.java]: Thought i had removed those. Must not have commited it.

File: code/implementation/docker/api/src/main/java/io/cattle/platform/docker/api/ContainerLogsActionHandler.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        HostApiAccess apiAccess = apiService.getAccess(host.getId(), CollectionUtils.asMap("exec", data));
+
+        if (apiAccess == null) {
+            return null;
+        }
+
+        StringBuilder url = new StringBuilder(CONSOLE_AGENT_SCHEME.get());
+        url.append("://").append(apiAccess.getHostname()).append(":").append(CONSOLE_AGENT_PORT.get());
+        url.append(CONSOLE_AGENT_PATH.get());
+        System.out.println("You hit the api\n\n\n\n\n\n\n\n");
```
Comment (ibuildthecloud) [inline code/implementation/docker/api/src/main/java/io/cattle/platform/docker/api/ContainerLogsActionHandler.java]: bad, bad, bad.  No sysout

File: code/implementation/docker/api/src/main/java/io/cattle/platform/docker/api/model/ContainerLogs.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package io.cattle.platform.docker.api.model;
+
+import io.github.ibuildthecloud.gdapi.annotation.Field;
+
+public interface ContainerLogs {
+
+    @Field(defaultValue = "true")
```
Comment (ibuildthecloud) [inline code/implementation/docker/api/src/main/java/io/cattle/platform/docker/api/model/ContainerLogs.java]: Either set the default on the java class or in the json, not both...

Comment (hibooboo2) [inline code/implementation/docker/api/src/main/java/io/cattle/platform/docker/api/model/ContainerLogs.java]: Ok I thought it needed to be in both wasnt sure as it is unclear

File: tests/integration/cattletest/core/test_container.py [hunk 9]
```diff
@@ -629,6 +629,32 @@ def test_container_exec(admin_client, sim_context):
...
+
+    assert resp.url is not None
+    assert resp.token is not None
+
+    jwt = _get_jwt(resp.token)
+
+    assert jwt['exec']['AttachStdin']
+    assert jwt['exec']['AttachStdout']
+    assert jwt['exec']['Tty']
+    assert jwt['exec']['Cmd'] == 'echo hello'
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_container.py]: Does this really makes sense???

Comment (hibooboo2) [inline tests/integration/cattletest/core/test_container.py]: I forgot to change that too sorry good catch Will fix it tomorrow i really need to be more cautious about clean up

File: code/implementation/docker/api/src/main/java/io/cattle/platform/docker/api/ContainerLogsActionHandler.java (hunk 7 above)
Comment (hibooboo2) [inline code/implementation/docker/api/src/main/java/io/cattle/platform/docker/api/ContainerLogsActionHandler.java]: Yeah i forgot to remove... :disappointed:

File: resources/content/schema/base/containerLogsInput.json (hunk 2 above)
Comment (hibooboo2) [inline resources/content/schema/base/containerLogsInput.json]: Addressed

File: resources/content/schema/user/user-auth.json (hunk 4 above)
Comment (hibooboo2) [inline resources/content/schema/user/user-auth.json]: But this is for the input not the output... modeled after line 81

File: code/iaas/model/src/main/resources/META-INF/cattle/core-model/spring-core-model-context.xml (hunk 1 above)
Comment (hibooboo2) [inline code/iaas/model/src/main/resources/META-INF/cattle/core-model/spring-core-model-context.xml]: This is needed otherwise you get a 404 for the logs action

Comment (ibuildthecloud) [inline code/iaas/model/src/main/resources/META-INF/cattle/core-model/spring-core-model-context.xml]: We need to discuss then, this shouldn't be needed

File: code/implementation/docker/api/src/main/resources/schema/base/container.json.d/docker.json [hunk 10]
```diff
@@ -22,10 +22,14 @@
...
         "execute": {
             "input": "containerExec",
-            "output": "containerExecOutput",
+            "output": "hostAccess",
             "attributes" : {
                 "state" : "running"
             }
+        },
+        "logs":{
+            "input":"containerLogsInput",
```
Comment (ibuildthecloud) [inline code/implementation/docker/api/src/main/resources/schema/base/container.json.d/docker.json]: type here should be containerLogs to match the java class.

File: resources/content/db/mysql/create_db_and_user_dev.sql [hunk 11]
```diff
@@ -1,8 +1,6 @@
-CREATE DATABASE cattle COLLATE = 'utf8_general_ci' CHARACTER SET = 'utf8';
+CREATE DATABASE IF NOT EXISTS cattle COLLATE = 'utf8_general_ci' CHARACTER SET = 'utf8';
```
Comment (ibuildthecloud) [inline resources/content/db/mysql/create_db_and_user_dev.sql]: Why did you change this file?

File: resources/content/schema/base/container.json [hunk 12]
```diff
@@ -187,5 +187,11 @@
             "type":"array[string]",
             "nullable": true
         }
+    },
```
Comment (ibuildthecloud) [inline resources/content/schema/base/container.json]: This seems redundant from docker.json.  I think this can be removed

File: resources/content/schema/base/containerLogsInput.json (hunk 2 above)
Comment (ibuildthecloud) [inline resources/content/schema/base/containerLogsInput.json]: Set default to true, not "true"

File: resources/content/schema/user/user-auth.json (hunk 4 above)
Comment (ibuildthecloud) [inline resources/content/schema/user/user-auth.json]: Really, this should be "containerLogs"

File: tests/integration/cattletest/core/test_container_logs.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+            found_ip = ip
+
+    assert found_ip is not None
+    assert found_ip.address is not None
+
+    return found_ip
+
+
+@if_docker
+def test_logs_container(admin_client):
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_container_logs.py]: If using docker always depend on the `docker_context` fixture.  This ensures that docker is properly setup.

File: code/iaas/model/src/main/resources/META-INF/cattle/core-model/spring-core-model-context.xml (hunk 1 above)
Comment (hibooboo2) [inline code/iaas/model/src/main/resources/META-INF/cattle/core-model/spring-core-model-context.xml]: Ok

File: resources/content/schema/base/containerLogsInput.json (hunk 2 above)
Comment (hibooboo2) [inline resources/content/schema/base/containerLogsInput.json]: Though i had guess not. Fixed it

File: resources/content/db/mysql/create_db_and_user_dev.sql (hunk 11 above)
Comment (hibooboo2) [inline resources/content/db/mysql/create_db_and_user_dev.sql]: I was using it for personal reasons to stop sql from erroring when the script is run if there is a database and users already made. I will take it out and submit a separate  pr with the change and an explanation as to why.

File: resources/content/schema/base/container.json (hunk 12 above)
Comment (hibooboo2) [inline resources/content/schema/base/container.json]: Thought it had to be in both places sorry. Hadn't caught the dups as when i did some of this i was still learning your frame work. You have multiple places to put things to make your magic happen.

Comment (hibooboo2) [inline resources/content/schema/base/container.json]: Left this one there removed the other.

**Comment by hibooboo2** [comment]
At 11:04 build was successful. I am now rebasing on master and re pushing.
